### PR TITLE
Fixing error thrown when unloading gun when there is no inventory available 

### LINKED ISSUE #29839: Unloading gun you're wielding when you have nowhere to store its ammo leads to an error
**Describe the bug**  
Unloading gun gave error then put a None item into player inventory.
With None item in inventory: Advanced inventory: Player Inventory->Vehicle Inventory: 'Move All' action crashed game.

**To Reproduce**  
Complete the 3rd? mission for the old guard rep in the refugee center
Get rewarded with Marshall badge and gun (plus other stuff)
My gun contained a single round of ammo (don't know if this is random)
Unload the gun.

**Versions and configuration(please complete the following information):

Version: 0.D-2038-g49c568e705 (Windows 10-64 Tiles - Custom build with VS2015 but with no code changes)
Mods: [ "dda", "aftershock", "filthy_morale", "no_npc_food", "growable-pots", "makeshift", "Medieval_Stuff", "More_Survival_Tools", "modular_turrets", "Salvaged_Robots", "mutant_npcs", "more_locations", "FujiStruct", "Urban_Development", "deoxymod", "blazemod", "Tanks", "no_flaming_weapons" ]

This is the only error in the debug log, not sure if it is related: ERROR : ..\src\player.cpp:8434 [item_reload_cost] cannot determine reload cost as none is neither ammo or magazine

**Expected Behavior**
Unloading should not give none item.
Should None be in inventory (as some sort of emergency placeholder) Interacting it should not cause crash

### LINKED ISSUE #47001: Unloading the a STANAG magazine causes the bullet to duplicate
# Describe the bug

When unloading a magazine with a few bullets inside and a full inventory you get prompted to drop the magazine, if you do, the bullets get duplicated, half inside the magazine and half in your hands. You get an error message when unloading the magazine
(edit: apparently it only works with a magazine that was naturally generated with only a few bullets. The bug doesn't work if you reload the magazine before or if the magazine has a bullet count bigger than one digit). 
(edit 2: it also works with shotguns and other magazines).

# Steps To Reproduce

1. Have a full inventory;
2. Hold a STANAG magazine in your hands;
3. Unload the magazine;
4. When prompted, drop the magazine.

# Expected behavior

The bullets get duplicated, with half of them inside the magazine and half in your hands. Unloading the magazine causes a error screen to appear.

# Screenshots

-magazine with bullets inside-

-unloading causes an error message to appear-

-getting prompted to drop the magazine-

-half of the bullets in my hand-

-other half of the bullets still inside the magazine-

# Versions and configuration
- OS: Windows
    - OS Version: 10.0 1909
- Game Version: 0.E-9181-g308bec1-dirty [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Bionic Professions [package_bionic_professions],
    Blaze Industries [blazeindustries],
    Blaze Industries - UndeadPeople Tileset [BlazeIndustriesUDP],
    C.R.I.T Expansion Mod [crt_expansion],
    C.R.I.T Expansion Mod - UndeadPeople Tileset [crt_expansionUDP],
    Magiclysm [magiclysm],
    Magiclysm - UndeadPeople Tileset [MagiclysmUDP],
    Aftershock [aftershock],
    Aftershock - UndeadPeople Tileset [AftershockUDP],
    Crazy Cataclysm [crazy_cataclysm],
    Crazy Cataclysm - UndeadPeople Tileset [CrazyCataclysmUDP],
    Mythical Martial Arts [MMA],
    Mythical Martial Arts - UndeadPeople Tileset [MMAUDP],
    Graphical Overmap [Graphical_Overmap],
    Graphical Overmap Magiclysm [Graphical_Overmap_Magiclysm],
    Fuji's Military Profession Pack [fuji_mpp],
    SpeedyDex [speedydex],
    Stats Through Kills [stats_through_kills],
    Stats Through Skills [StatsThroughSkills],
    Bionic Slots [cbm_slots]
]

# Additional context
I forgot to take a screenshot of the error before pressing to ignore...
[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/5863068/debug.log)

### LINKED ISSUE #48842: Thread duplication
# Describe the bug

Trying to (U)nload a partial amount of Thread (and Nomex/Sinew) from a (w)ielded Tailor's Kit with a full inventory gives the error
```
DEBUG    : Tried removing item from object which did not contain it

 FUNCTION : item visitable::remove_item(item&)
 FILE     : src/visitable.cpp
 LINE     : 496
```

# Steps To Reproduce

1. Take off all clothing
2. Load ~100 or so thread into a Tailor's Kit
3. (w)ield the Tailor's Kit
4. Attempt to (U)nload the Tailor's Kit

# Expected behavior

Thread is safely removed without any duplication.

# Versions and configuration

 - OS: Windows 10
 - Game Version: 0.E-10395-g4525b4f
 - Graphics version: Tiles
 - Ingame language: English
 - Mods loaded: none.

# Additional context

Brand new debug world in the correct conditions to trigger the duplication.

[Test.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/6462656/Test.zip)

This MIGHT have been fixed by #47885 but I'm having some issues updating versions at the moment and this is occurring outside of AIM, and has a debug message not mentioned in the other issues of that type.

### PR DESCRIPTION
**Pull request by lancewf**
#### Summary
BugFixes "Unloading gun you're wielding when you have nowhere to store its ammo leads to an error"

#### Purpose of change

Fixes #29839
Fixes #47001
Fixes #48842

To reproduce the bug: 
1. Spawn a ruger_redhawk revolver
2. Unload the weapon with "U"
3. Drop the .44 bullets
2. Drop all other items, but the revolver
4. Load one bullet in the revolver with "r"
5. Unload the weapon again with "U". After this action (before the code change) the player would be promoted to drop their weapon. Then when dropping the weapon an error would be displayed. 

This problem happens with shotguns or any other gun without a magazine that takes more than one round. 

#### Describe the solution

The problem only happened when a player only had inventory space for the unloaded bullets in the gun they are unloading. The game thinks it has inventory space to store the bullets, but it does not. Then it tries to drop the gun and store the bullets in the gun at the same time. The solution is to not check for inventory space in the gun that is being unloaded. Then no inventory space will be found and the bullets will be dropped on the ground. 

#### Describe alternatives you've considered

An alternative is the https://github.com/CleverRaven/Cataclysm-DDA/pull/48895 PR that I have confirmed solves #29839 also. 

Although they both solve the issue they are solving different problems in the code. This PR excludes the item being unloaded's inventory volume from being counted to placed the unloaded items. The other PR fixes how we check if the wielded item is being unloaded. Both changes are needed, but either will fix the three issues. 

Below is the section of code that both PRs fix. Line 12807 is where this PR fixes the problem. This line is checking if there is space to store the unloaded items. Line 12812 is where the other PR fixes the problem. This line checks if the item unloaded is in the item wielded. If Line 12807 is fixed (this PR) then the code will not get to line 12812 (other PR) because it will find that there is not enough space to store the unloaded item and it will drop it or place it in the vehicle (line 12808). If line 12807 is not fixed and line 12812 (other PR) is, it will get to line 12812 and set `allow_wield` to false because the unloaded item is part of the wielded item. Then the items will be dropped from line 12814. 

https://github.com/CleverRaven/Cataclysm-DDA/blob/38a368d22e2fac321212e8b88f67b5891ef25474/src/character.cpp#L12801-L12829

We need to fix line 12807 (this PR) because it will make the code less confusing to get to line 12812 when there is not enough volume to place the unloaded item. How I read this code is "Do we have enough volume", "Do we have too much weight", then place the item. 

We need to fix line 12812 (other PR) because it is not working correctly. It is checking if the unloaded item is in the weapon which it is and currently it is returning that it is not. 

#### Testing

There are unit tests created for the problem. There were no problems found when testing this change with other types of weapons.

Manual testing with the wielded item below. With full and partially loaded items. 
* sw_610
* sw_500
* m2browning_sawn
* ruger_redhawk
* sw629
* ruger_lcr_38
* sw_619
* sew kit
* STANGA magazine

### CODE REVIEW

File: src/character.cpp [hunk 1]
```diff
@@ -3680,9 +3680,9 @@ units::mass Character::weight_capacity() const
     return ret;
 }
 
-bool Character::can_pickVolume( const item &it, bool ) const
+bool Character::can_pickVolume( const item &it, bool, const item *avoid ) const
 {
-    if( weapon.can_contain( it ) ) {
+    if( weapon.can_contain( it ) && ( avoid == nullptr || &weapon != avoid )) {
```
Comment (pehamm) [inline src/character.cpp:3685]:
```suggestion
    if( weapon.can_contain( it ) && ( avoid == nullptr || &weapon != avoid ) ) {
```
I believe this is why the astyle check fails

File: src/activity_actor.cpp [hunk 2]
```diff
@@ -1646,7 +1646,7 @@ void unload_activity_actor::unload( Character &who, item_location &target )
         if( contained->ammo_type() == ammotype( "plutonium" ) ) {
             contained->charges /= PLUTONIUM_CHARGES;
         }
-        if( who.as_player()->add_or_drop_with_msg( *contained, true ) ) {
+        if( who.as_player()->add_or_drop_with_msg( *contained, true, &it ) ) {
```
Comment (lancewf) [inline src/activity_actor.cpp:1649]: This avoids counting the volume of the item that is being unloaded as space the unloaded items can go to.

File: src/character.cpp [hunk 3]
```diff
@@ -3680,9 +3680,9 @@ units::mass Character::weight_capacity() const
...
 }
 
-bool Character::can_pickVolume( const item &it, bool ) const
+bool Character::can_pickVolume( const item &it, bool, const item *avoid ) const
 {
-    if( weapon.can_contain( it ) ) {
+    if( weapon.can_contain( it ) && ( avoid == nullptr || &weapon != avoid ) ) {
         return true;
     }
     for( const item &w : worn ) {
```
Comment (lancewf) [inline src/character.cpp:3688]: We might want to add the avoid check here also, but I could not find a case with non-weapons that cause a problem.

File: tests/unload_naked_test.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    item_location revo_loc( dummy, &player_character.weapon );
+    player_character.moves = 100;
+    REQUIRE( player_character.unload( revo_loc ) );
+    player_character.activity.do_turn( player_character );
+
+    // No bullets in wielded gun
+    CHECK( player_character.weapon.ammo_remaining() == 0 );
+
+    // No bullets in inventory
+    const std::vector<item *> bullets = dummy.items_with( []( const item& item ) {
```
Comment (actual-nh) [inline tests/unload_naked_test.cpp:49]:
```suggestion
    const std::vector<item *> bullets = dummy.items_with( []( const item & item ) {
```
Astyle.

File: tests/unload_naked_test.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+TEST_CASE( "unload_revolver_naked_fully_loaded", "[unload][nonmagzine]" )
+{
+    clear_avatar();
+    clear_map();
+
+    Character &dummy = get_player_character();
+    avatar &player_character = get_avatar();
+
+    // revolver fully loaded
```
Comment (actual-nh) [inline tests/unload_naked_test.cpp:63]:
```suggestion
    // revolver fully loaded
```
Space at end (astyle), as far as I can figure out.

File: tests/unload_naked_test.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    item_location revo_loc( dummy, &player_character.weapon );
+    player_character.moves = 100;
+    REQUIRE( player_character.unload( revo_loc ) );
+    player_character.activity.do_turn( player_character );
+
+    // No bullets in wielded gun
+    CHECK( player_character.weapon.ammo_remaining() == 0 );
+
+    // No bullets in inventory
+    const std::vector<item *> bullets = dummy.items_with( []( const item& item ) {
```
Comment (actual-nh) [inline tests/unload_naked_test.cpp:84]:
```suggestion
    const std::vector<item *> bullets = dummy.items_with( []( const item & item ) {
```
Astyle.

**Review by actual-nh — COMMENTED** [review]
These will hopefully fix the astyle objections.

**Comment by lancewf** [comment]
@actual-nh Thank you for fixing the astyle check. 

I am not sure how to read the problem with the Clang-tidy error. I am not even sure which file is the problem, but I am assuming it is the "unload_naked_test.cpp".

**Comment by pehamm** [comment]
You can ignore the clang-tidy error, it is caused by an problem with a python library needed for the workflow and should be fixed once #49000 is merged.

**Comment by RoyBerube** [comment]
This looks very similar to what is being addressed in PR #48895, although the code fixes it in a different way.

**Comment by actual-nh** [comment]
> This looks very similar to what is being addressed in PR #48895, although the code fixes it in a different way.

Perhaps a combination of the approaches? The tests from this PR are definitely a good thing to include.

**Comment by lancewf** [comment]
> This looks very similar to what is being addressed in PR #48895, although the code fixes it in a different way.

I confirmed that #48895 does fix this issue. Although they both solve the issue they are solving different problems in the code. This PR excludes the volume from the item being unloaded for a place to put the items. The other PR fixes how we check if the wielded item is being unloaded. To me, it looks like both changes are needed. 

I am testing them both together now.

File: src/character.cpp [hunk 7]
```diff
@@ -12807,12 +12807,21 @@ bool Character::add_or_drop_with_msg( item &it, const bool /*unloading*/, const
...
+        bool wielded_has_it = false;
+        // Cannot use weapon.has_item(it) because it skips any pockets that
+        // are not containers such as magazines and magazine wells.
+        for( const item *scan_contents : weapon.contents.all_items_top() ) {
+            if( scan_contents == &it ) {
+                wielded_has_it = true;
+                break;
+            }
+        }
+        const bool allow_wield = !wielded_has_it && weapon.magazine_current() != &it;
```
Comment (lancewf) [inline src/character.cpp:12824]: This section of code was cherry-picked from #48895. I think both these changes are needed for this fix.
If we want these changes we can merge #48895 first or only merge this PR.

**Comment by lancewf** [comment]
I cherry-picked the commit from #48895 (I hope this is ok @RoyBerube) to combine the two PRs. There were no problems found when testing the combined PRs with a few guns, a magazine, and the sewing kit.

**Comment by actual-nh** [comment]
@lancewf: #48895 is now in; do you need to rebase this PR?

**Comment by lancewf** [comment]
@actual-nh We should not need to because the commit (60976a602b69307a8a40367594665c601e7d3db7) from #48895 was pulled into this branch.

**Pull request merged**